scripts/apps/contacts/services/ContactsService.ts

```diff
@@ -140,9 +164,37 @@ export class ContactsService {
         return this.api.save('contacts', {public: isPublic});
     }
 
-    save(contact, data) {
-        return this.api.save('contacts', contact, data);
+    save(contact, updates) {
+        return this.api.save('contacts', contact, this.convertForServer(updates))
+            .then(this.convertForClient);
+    }
+
+    convertForServer(contact: IContact) {
+        if (!contact || typeof contact.contact_type === 'string') {
+            return contact;
+        } else if (contact.contact_type.qcode) {
+            contact.contact_type = contact.contact_type.qcode;
+        }
+
+        return contact;
+    }
+
+    convertForClient(response: IContact) {
+        if (!response) {
+            return response;
+        }
+
+        const contact: IContact = replaceUrls(response);
+
+        if (typeof contact.contact_type === 'string' && get(this.metadata, 'values.contact_type.length', 0) > 0) {
```

Comment:
What's the purpose of the conversion? Why doesn't `contact_type` from the server work?

Reply:
The server stores `contact_type` as a string, where as we want to use the object version of the value in the client `IContactType` (which includes `qcode`, `name` and `assignable` fields)

Reply:
Wouldn't it be cleaner to keep the entity exactly as it is on the server and load additional data outside of that data structure?

```
export interface IContact {
    _id: string;
    contact_type?: string;
}

const contacts: Array<IContact> = getContacts();

contacts.forEach((contact) => {
    const contactTypeObj = metadata[contact.contact_type];

    console.log(contactTypeObj.name);
});
```

We do a similar thing with users frequently. If we need to get a display name of a user, we don't patch the object which has the user as an ID string. We load full user objects separately and can then retrieve these objects by IDs.